cdc/log.cc

```diff
@@ -722,15 +824,57 @@ class transformer final {
                         res.set_cell(log_ck, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
                     }
 
-                    if (has_pirow) {
-                        value = get_preimage_col_value(cdef, pirow);
+                    bytes_opt prev;
 
+                    if (has_pirow) {
+                        prev = get_preimage_col_value(cdef, pirow);
                         assert(std::addressof(res.partition().clustered_row(*_log_schema, *pikey)) != std::addressof(res.partition().clustered_row(*_log_schema, log_ck)));
                         assert(pikey->explode() != log_ck.explode());
-                        res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
+                        res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, *prev, _cdc_ttl_opt));
+                    }
+
+                    if (postimage) {
+                        // keep track of actually assigning this already
+                        columns_assigned.emplace(id);
+                        // don't merge with pre-image iff column delete
+                        if (is_column_delete) {
+                            prev = std::nullopt;
+                        }
+                        if (cdef.is_atomic() && !is_column_delete && value) {
+                            res.set_cell(*poikey, *dst, atomic_cell::make_live(*dst->type, ts, *value, _cdc_ttl_opt));
+                        } else if (!cdef.is_atomic() && (value || (deleted_elements && prev))) {
+                            auto v = visit(*cdef.type, make_visitor(
```

Comment:
Would it be possible to reduce code duplication by using a generic lambda visitor here? That would require having merge with overloads for collection_type_impl, set_type_impl and user_type_impl instead of having merege_maps, merge_sets and merge_udt.

Reply:
Not sure what you are getting at?

Reply:
```
auto v = visit(*cdef.type, make_visitor([&] (const auto& type) -> bytes {
  return merge(type, prev, value, deleted_elements);
}));

and then create merge for each type (collection_type_impl, set_type_impl, user_type_impl).
```

Reply:
This could be
``make_visitor([&] (const auto& type) -> bytes { return merge(*type.name_comparator(), prev, value, deleted_elements); })``
where `merge` is overloaded in the first parameter (type). See also my comment below `merge_maps`.

But there's a problem that we have the `const abstract_type&` branch which throws an error. But maybe we could make two branches: one generic like shown in the example above for collections and UDTs, and one specific for `const abstract_type&`. Dunno if the compiler will handle this monster.

Reply:
With all the template<..> lines added it will not cut down the size of the file. I enjoy overloads as much as the other guy, but I think it does not really help here (it is 8 lines)

Reply:
This wouldn't add any additional `template<...>` lines, you would just need to change name of `merge_maps` to `merge`, name of `merge_sets` to `merge` etc.

Also in this case it's less about number of lines, more about kosherness :P